src/liballoc/collections/btree/node.rs

```diff
@@ -515,20 +515,25 @@ impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Immut<'a>, K, V, Type> {
         // Instead, we create a slice pointing into the node whenever possible.
         // We can sometimes do this even for the shared root, as the slice will be
         // empty.  We cannot *always* do this because if the type is too highly
-        // aligned, the offset of `keys` in a "full node" might be outside the bounds
-        // of the header!  So we do an alignment check first, that will be
-        // evaluated at compile-time, and only do any run-time check in the rare case
-        // that the alignment is very big.
-        if mem::align_of::<K>() > mem::align_of::<LeafNode<(), ()>>() && self.is_shared_root() {
+        // aligned, the offset of a correctly typed `keys_start` might be outside
+        // the bounds of the allocated header!  So we do an alignment check first,
+        // that will be evaluated at compile-time, and only do any run-time check
+        // in the rare case that the alignment is very big.
+        // We check size in addition to alignment, because we want to be absolutely
+        // sure about a field at the back of the `repr(C) struct` header.
+        if (mem::align_of::<NodeHeader<K, V, K>>() > mem::align_of::<NodeHeader<K, V>>()
+            || mem::size_of::<NodeHeader<K, V, K>>() > mem::size_of::<NodeHeader<K, V>>())
```

Comment:
```suggestion
            || mem::size_of::<NodeHeader<K, V, K>>() != mem::size_of::<NodeHeader<K, V>>())
```
A smaller size would also totally ruin the computations here.

Reply:
I guess so, even though even I feel confident that changing a type to something that imposes nothing cannot stretch the struct. But surely you mean the same for size_of as well?

Reply:
I meant only `size_of`.

Reply:
I meant to write, and meanwhile changed it for, `align_of` as well. But I see your point now: If the allocated header has coarser alignment than what we need, transmuting is still safe.

Reply:
So now it's there (in `into_key_slice_mut` too)